From 503b9c111a31da738e4bd599ef15fc0b0c6dbd7e Mon Sep 17 00:00:00 2001
From: Pratham Patel <ppatel@ciq.com>
Date: Fri, 8 Aug 2025 14:39:48 +0530
Subject: [PATCH] scsi: libfc: Fix use after free in fc_exch_abts_resp()

jira VULN-49947
cve CVE-2022-49114
commit-author Jianglei Nie <niejianglei2021@163.com>
commit 271add11994ba1a334859069367e04d2be2ebdd4

fc_exch_release(ep) will decrease the ep's reference count. When the
reference count reaches zero, it is freed. But ep is still used in the
following code, which will lead to a use after free.

Return after the fc_exch_release() call to avoid use after free.

Link: https://lore.kernel.org/r/20220303015115.459778-1-niejianglei2021@163.com
	Reviewed-by: Hannes Reinecke <hare@suse.de>
	Signed-off-by: Jianglei Nie <niejianglei2021@163.com>
	Signed-off-by: Martin K. Petersen <martin.petersen@oracle.com>
(cherry picked from commit 271add11994ba1a334859069367e04d2be2ebdd4)
	Signed-off-by: Pratham Patel <ppatel@ciq.com>
---
 drivers/scsi/libfc/fc_exch.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/scsi/libfc/fc_exch.c b/drivers/scsi/libfc/fc_exch.c
index 317cc9b0dced..ef30d7563cad 100644
--- a/drivers/scsi/libfc/fc_exch.c
+++ b/drivers/scsi/libfc/fc_exch.c
@@ -1701,6 +1701,7 @@ static void fc_exch_abts_resp(struct fc_exch *ep, struct fc_frame *fp)
 	if (cancel_delayed_work_sync(&ep->timeout_work)) {
 		FC_EXCH_DBG(ep, "Exchange timer canceled due to ABTS response\n");
 		fc_exch_release(ep);	/* release from pending timer hold */
+		return;
 	}
 
 	spin_lock_bh(&ep->ex_lock);
-- 
2.47.3

